src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsFileWatcher.cs

```diff
@@ -0,0 +1,180 @@
+
+            // we don't care about the difference between types of inputs, so we just construct one hashset for fast comparisons later
+            var allFiles = new HashSet<string>(StringComparers.Paths);
+            allFiles.AddRange(designTimeInputs.Inputs);
+            allFiles.AddRange(designTimeInputs.SharedInputs);
+
+            // Remove any files we're watching that we don't care about any more
+            foreach (string file in _fileWatcherCookies.Keys.ToArray())
+            {
+                if (!allFiles.Contains(file))
+                {
+                    ImmutableInterlocked.TryRemove(ref _fileWatcherCookies, file, out uint cookie);
+                    await vsAsyncFileChangeEx.UnadviseFileChangeAsync(cookie);
+                }
+            }
+
+            // Now watch and output files that are new
+            foreach (string file in allFiles)
+            {
+                if (!_fileWatcherCookies.ContainsKey(file))
+                {
+                    // We don't care about delete and add here, as they come through data flow, plus they are really bouncy - every file change is a Time, Del and Add event)
+                    uint cookie = await vsAsyncFileChangeEx.AdviseFileChangeAsync(file, _VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size, sink: this);
+
+                    ImmutableInterlocked.TryAdd(ref _fileWatcherCookies, file, cookie);
+
+                    // Advise of an addition now
+                    PostToOutput(file);
+                }
+            }
+        }
+
+        private void PostToOutput(string file)
+        {
+            _version++;
+            ImmutableDictionary<NamedIdentity, IComparable> dataSources = ImmutableDictionary<NamedIdentity, IComparable>.Empty.Add(DataSourceKey, DataSourceVersion);
+
+            _broadcastBlock.Post(new ProjectVersionedValue<string>(file, dataSources));
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _broadcastBlock?.Complete();
+                _disposables.Dispose();
+
+                _threadingService.ExecuteSynchronously(async () =>
+                {
+                    IVsAsyncFileChangeEx vsAsyncFileChangeEx = await _fileChangeService.GetValueAsync();
+
+                    foreach (uint cookie in _fileWatcherCookies.Values)
+                    {
+                        await vsAsyncFileChangeEx.UnadviseFileChangeAsync(cookie);
+                    }
+                });
+            }
+
+            base.Dispose(disposing);
+        }
+
+        public int FilesChanged(uint cChanges, string[] rgpszFile, uint[] rggrfChange)
+        {
+            for (int i = 0; i < cChanges; i++)
+            {
+                PostToOutput(rgpszFile[i]);
+            }
+
+            return HResult.OK;
+        }
+
+        public int DirectoryChanged(string pszDirectory)
+        {
+            throw new NotImplementedException();
```

Comment:
Return `HResult.NotImplemented` and below, you can't throw from a COM interface as you are breaking the contract and managed consumers will need to handle both HRs and exceptions.

Reply:
Thanks